mkosi

```diff
@@ -268,13 +269,27 @@ def init_namespace(args):
 def setup_workspace(args):
     print_step("Setting up temporary workspace.")
     if args.output_format in (OutputFormat.directory, OutputFormat.subvolume):
-        d = tempfile.TemporaryDirectory(dir=os.path.dirname(args.output), prefix='.mkosi-')
+        d = Workspace(dir=os.path.dirname(args.output), prefix='.mkosi-')
     else:
-        d = tempfile.TemporaryDirectory(dir='/var/tmp', prefix='mkosi-')
+        d = Workspace(dir='/var/tmp', prefix='mkosi-')
 
     print_step("Temporary workspace in " + d.name + " is now set up.")
     return d
 
+class Workspace(object):
+    def __init__(self, suffix=None, prefix=None, dir=None):
+        self._tempdir = tempfile.TemporaryDirectory(suffix=suffix, prefix=prefix, dir=dir)
+        weakref.finalize(self, self._cleanup, self._tempdir)
```

Comment:
Hmm, can you explain why `weakref` is used instead of just defining a `__del__` method?

Reply:
(sorry for the late review)

Reply:
I think I started from the tempfile.TemporaryDirectory implementation, but I removed the `__enter__` and `__exit__` methods as they were not used. `__del__` might indeed be enough.

Reply:
I think the nicest way to implement this would be subclass TemporaryDirectory and only override .cleanup() to call `empty_directory()` first.